tls/s2n_early_data.c

```diff
@@ -167,6 +167,18 @@ S2N_RESULT s2n_early_data_get_server_max_size(struct s2n_connection *conn, uint3
     return S2N_RESULT_OK;
 }
 
+int s2n_connection_set_server_early_data_context(struct s2n_connection *conn, const uint8_t *context, uint16_t context_size)
+{
+    POSIX_ENSURE_REF(conn);
+    if (context_size > 0) {
+        POSIX_ENSURE_REF(context);
+    }
+
+    POSIX_GUARD(s2n_realloc(&conn->server_early_data_context, context_size));
+    POSIX_CHECKED_MEMCPY(conn->server_early_data_context.data, context, context_size);
```

Comment:
Note to readers:

This code is safe even when `context_size == 0` and `context` is not valid because the underlying implementation of this macro in s2n checks that `context_size !=0` prior to calling the underlying `memcpy` method.

In general, both the source and destination pointers for `memcpy` **must** be valid even when `len == 0` and if they are not then it may be a correctness (and potential security) issue. Fortunately, s2n's macros keep us safe here.